packages/labs/context/src/lib/controllers/context-provider.ts

```diff
@@ -83,19 +104,52 @@ export class ContextProvider<T extends Context<unknown, unknown>>
     // The check on composedPath (as opposed to ev.target) is to cover cases
     // where the consumer is in the shadowDom of the provider (in which case,
     // event.target === this.host because of event retargeting).
-    if (ev.context !== this.context || ev.composedPath()[0] === this.host) {
+    const consumerHost = ev.composedPath()[0] as Element;
+    if (ev.context !== this.context || consumerHost === this.host) {
       return;
     }
     ev.stopPropagation();
-    this.addCallback(ev.callback, ev.subscribe);
+    this.addCallback(ev.callback, ev.subscribe, consumerHost);
+  };
+
+  public onProviderRequest = (
+    ev: ContextProviderEvent<Context<unknown, unknown>>
+  ): void => {
+    // Ignore events when the context doesn't match.
+    // Also, in case an element is a consumer AND a provider
+    // of the same context, we want to avoid the element to self-register.
+    // The check on composedPath (as opposed to ev.target) is to cover cases
+    // where the consumer is in the shadowDom of the provider (in which case,
+    // event.target === this.host because of event retargeting).
+    const childProviderHost = ev.composedPath()[0] as Element;
+    if (
+      !ev.canTakeOverSubscriptions ||
+      ev.context !== this.context ||
+      childProviderHost === this.host
+    ) {
+      return;
+    }
+    const ourEv = ev as ContextProviderEvent<T>;
+    const movedSubscriptions = this.moveSubscriptionsFor(childProviderHost);
+    if (movedSubscriptions === undefined) {
+      return;
+    }
+    (ourEv.movedSubscriptions ??= []).push(...movedSubscriptions);
   };
 
   private attachListeners() {
     this.host.addEventListener('context-request', this.onContextRequest);
+    this.host.addEventListener('context-provider', this.onProviderRequest);
   }
 
   hostConnected(): void {
     // emit an event to signal a provider is available for this context
-    this.host.dispatchEvent(new ContextProviderEvent(this.context));
+    const providerEvent = new ContextProviderEvent(this.context);
+    this.host.dispatchEvent(providerEvent);
+    if (providerEvent.movedSubscriptions !== undefined) {
```

Comment:
Nit & no action required: Is it too much to add a comment here that `movedSubscriptions` may be populated by an ancestor provider & they may need to be moved to this provider?